### PR TITLE
[Clang][Interp] Fix the location of uninitialized base warning

### PR DESCRIPTION
**Pull request by yronglin**
Fix the location of `diag::note_constexpr_uninitialized_base`, make it same as current interpreter.
This PR does not print type name with namespacethat was used to improve the current interpreter's type dump of base class type.

### CODE REVIEW

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 1]
```diff
@@ -122,22 +122,20 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
...
-        S.FFDiag(BasePtr.getDeclDesc()->asDecl()->getLocation(),
-                 diag::note_constexpr_uninitialized_base)
-            << B.Desc->getType();
-      else
-        S.FFDiag(BasePtr.getDeclDesc()->asExpr()->getExprLoc(),
-                 diag::note_constexpr_uninitialized_base)
-            << B.Desc->getType();
-
+      SourceLocation TypeBeginLoc = BS.getBaseTypeLoc();
+      SourceRange Range(TypeBeginLoc, BS.getEndLoc());
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:132]: Is this different than `BS.getSourceRange()`?

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 2]
```diff
@@ -122,22 +122,20 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
   }
 
   // Check Fields in all bases
-  for (const Record::Base &B : R->bases()) {
+  unsigned BaseIndex = 0;
+  const CXXRecordDecl *CD = dyn_cast<CXXRecordDecl>(R->getDecl());
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:126]:
```suggestion
  const auto *CD = dyn_cast<CXXRecordDecl>(R->getDecl());
```

Comment (tbaederr) [reply clang/lib/AST/Interp/EvaluationResult.cpp:126]: This either needs to be a `cast<>` or a null check. And I think it should be a null check.

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 3]
```diff
@@ -122,22 +122,20 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
...
-            << B.Desc->getType();
-
+      SourceLocation TypeBeginLoc = BS.getBaseTypeLoc();
+      SourceRange Range(TypeBeginLoc, BS.getEndLoc());
+      S.FFDiag(TypeBeginLoc, diag::note_constexpr_uninitialized_base)
+          << B.Desc->getType() << Range;
       return false;
     }
     Result &= CheckFieldsInitialized(S, Loc, P, B.R);
+    BaseIndex++;
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:138]:
```suggestion
    ++BaseIndex;
```

**Review by tbaederr — COMMENTED** [review]
Screenshot from 2024-07-27 07-04-02

The three regressions all have similar stack traces:
```
 #5 0x000000000bba9f4c clang::Redeclarable<clang::TagDecl>::getFirstDecl() /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/Redeclarable.h:216:38
 #6 0x000000000bba9cd5 clang::Redeclarable<clang::TagDecl>::getMostRecentDecl() /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/Redeclarable.h:227:12
 #7 0x000000000bba9a69 clang::RecordDecl::getMostRecentDecl() /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/Decl.h:4168:59
 #8 0x000000000bba99b5 clang::CXXRecordDecl::getMostRecentDecl() /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:542:46
 #9 0x000000000bba9995 clang::CXXRecordDecl::getMostRecentDecl() const /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:546:5
#10 0x000000000bba9969 clang::CXXRecordDecl::dataPtr() const /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:458:5
#11 0x000000000bba9885 clang::CXXRecordDecl::data() const /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:463:11
#12 0x000000000bc92be5 clang::CXXRecordDecl::bases_begin() const /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:627:58
#13 0x000000000bc7a6a9 clang::CXXRecordDecl::bases() const /home/fedora/interp-tests/llvm-project/clang/include/clang/AST/DeclCXX.h:623:35
#14 0x00000000127e9d9f clang::interp::CheckFieldsInitialized(clang::interp::InterpState&, clang::SourceLocation, clang::interp::Pointer const&, clang::interp::Record const*) /home/fedora/interp-tests/llvm-project/clang/lib/AST/Interp/EvaluationResult.cpp:127:41
```
So I think it's cases where the decl is no CXXRecordDecl. 
This patch also drops the handling where the pointer descriptor was created for an expression - this might be relevant for those tests or not.

File: clang/lib/AST/Interp/EvaluationResult.cpp (hunk 2 above)
Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Done.

File: clang/lib/AST/Interp/EvaluationResult.cpp (hunk 1 above)
Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Yeah, they are same, I've dumped 2 source ranges in my local.

File: clang/lib/AST/Interp/EvaluationResult.cpp (hunk 3 above)
Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Done.

File: clang/test/AST/Interp/constexpr-subobj-initialization.cpp [hunk 4]
```diff
@@ -14,33 +14,29 @@ struct DelBase {
   constexpr DelBase() = delete; // expected-note {{'DelBase' has been explicitly marked deleted here}}
```
[the remark this answers is not in the document]
Comment (yronglin) [reply clang/test/AST/Interp/constexpr-subobj-initialization.cpp:12]: Removed 2nd bullet.

**Comment by tbaederr** [comment]
CC @AaronBallman @Sirraide @shafik Since this adds new API to `CXXBaseSpecifier`.

**Review by Sirraide — CHANGES_REQUESTED** [review]
`CXXRecordDecl` already exposes `bases()` and `vbases()`, I don’t think we need this new API for accessing a base by index. This pr looks like it should just be using `llvm::enumerate`.

**Comment by yronglin** [comment]
> `CXXRecordDecl` already exposes `bases()` and `vbases()`, I don’t think we need this new API for accessing a base by index. This pr looks like it should just be using `llvm::enumerate`.

Thanks for the review! I've tried to use `llvm::enumerate(clang::interp::Record->bases(), clang::CXXRecordDecl->bases())`, but clang::CXXRecordDecl may be null in some case(clang::interp::Record was not a CXXRecordDecl). IIUC, if `clang::CXXRecordDecl` is null, we need only traverse `clang::interp::Record->bases()`.  I've removed these two APIs in `CXXRecordDecl`, and use `std::next` instead, WDYT?

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 5]
```diff
@@ -122,19 +123,22 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
...
+  for (unsigned I = 0, E = R->getNumBases(); I != E; ++I) {
+    const auto &B = *R->getBase(I);
     Pointer P = BasePtr.atField(B.Offset);
     if (!P.isInitialized()) {
       const Descriptor *Desc = BasePtr.getDeclDesc();
-      if (Desc->asDecl())
-        S.FFDiag(BasePtr.getDeclDesc()->asDecl()->getLocation(),
-                 diag::note_constexpr_uninitialized_base)
+      if (const auto *CD = dyn_cast_or_null<CXXRecordDecl>(R->getDecl())) {
+        const auto &BS = *std::next(CD->bases_begin(), I);
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:132]: Does `CD->bases()[I]` not work?

Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp:132]: Yeah, the compiler says "Type 'clang::CXXRecordDecl::base_class_const_range' (aka 'iterator_range<const clang::CXXBaseSpecifier *>') does not provide a subscript operator".

**Comment by Sirraide** [comment]
> I've removed these two APIs in `CXXRecordDecl`, and use `std::next` instead, WDYT?

Makes sense; that’s more or less what I had in mind too.

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 6]
```diff
@@ -122,19 +123,22 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
   }
 
   // Check Fields in all bases
-  for (const Record::Base &B : R->bases()) {
+  for (unsigned I = 0, E = R->getNumBases(); I != E; ++I) {
+    const auto &B = *R->getBase(I);
```
Comment (Sirraide) [inline clang/lib/AST/Interp/EvaluationResult.cpp:127]:
```suggestion
  for (auto [I, B] : llvm::enumerate(R->bases())) {
```
Maybe I wasn’t too clear about this, but what I had in mind was using `llvm::enumerate` here and then `std::next` below (which you already did). Is there a reason why this doesn’t work that I’m missing here?

Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Thanks, it's make sense.

**Review by Sirraide — APPROVED** [review]
Alright, that addresses all the changes I was concerned about, so this lgtm, but I’ll leave it to  @tbaederr to approve this for good since I’m not really familiar w/ the interpreter...

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 7]
```diff
@@ -122,19 +124,21 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
...
-                 diag::note_constexpr_uninitialized_base)
+      if (const auto *CD = dyn_cast_or_null<CXXRecordDecl>(R->getDecl())) {
+        const auto &BS = *std::next(CD->bases_begin(), I);
+        S.FFDiag(BS.getBaseTypeLoc(), diag::note_constexpr_uninitialized_base)
+            << B.Desc->getType() << BS.getSourceRange();
+      } else {
+        SourceLocation Loc =
+            Desc->asDecl() ? BasePtr.getDeclDesc()->asDecl()->getLocation()
+                           : BasePtr.getDeclDesc()->asExpr()->getExprLoc();
+        S.FFDiag(Loc, diag::note_constexpr_uninitialized_base)
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:139]:
```suggestion
        S.FFDiag(Desc->getLocation(), diag::note_constexpr_uninitialized_base)
```

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 8]
```diff
@@ -122,19 +124,21 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
...
   // Check Fields in all bases
-  for (const Record::Base &B : R->bases()) {
+  for (auto [I, B] : llvm::enumerate(R->bases())) {
     Pointer P = BasePtr.atField(B.Offset);
     if (!P.isInitialized()) {
       const Descriptor *Desc = BasePtr.getDeclDesc();
-      if (Desc->asDecl())
-        S.FFDiag(BasePtr.getDeclDesc()->asDecl()->getLocation(),
-                 diag::note_constexpr_uninitialized_base)
+      if (const auto *CD = dyn_cast_or_null<CXXRecordDecl>(R->getDecl())) {
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:131]:
```suggestion
      if (const auto *CD = dyn_cast_if_present<CXXRecordDecl>(R->getDecl())) {
```

Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Done.

File: clang/lib/AST/Interp/EvaluationResult.cpp (hunk 7 above)
Comment (yronglin) [reply clang/lib/AST/Interp/EvaluationResult.cpp]: Done.

File: clang/lib/AST/Interp/EvaluationResult.cpp [hunk 9]
```diff
@@ -122,19 +124,18 @@ static bool CheckFieldsInitialized(InterpState &S, SourceLocation Loc,
   }
 
   // Check Fields in all bases
-  for (const Record::Base &B : R->bases()) {
+  for (auto [I, B] : llvm::enumerate(R->bases())) {
```
Comment (tbaederr) [inline clang/lib/AST/Interp/EvaluationResult.cpp:127]: Does `auto &[I, B]` work here? What about `const auto &`?

Comment (Sirraide) [reply clang/lib/AST/Interp/EvaluationResult.cpp:127]: I believe `llvm::enumerate` acts like `std::views::enumerate` and iterates by reference, so you’re supposed to just do `auto [...]`; at least that’s what I recall from the documentation.

**Review by tbaederr — APPROVED** [review]

**Comment by yronglin** [comment]
@tbaederr @Sirraide Thanks for your review!

**Pull request merged**